Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/DriverReferral.hs

```diff
@@ -59,3 +69,46 @@ createDriverReferral (driverId, merchantId) isDashboard ReferralLinkReq {..} = d
             driverId = cast driverId,
             linkedAt = now
           }
+
+generateReferralCode ::
+  ( HasCacheConfig r,
+    Redis.HedisFlow m r,
+    MonadFlow m,
+    EsqDBReplicaFlow m r,
+    EsqDBFlow m r,
+    MonadTime m
+  ) =>
+  (Id SP.Person, Id DM.Merchant) ->
+  m GenerateReferralCodeRes
+generateReferralCode (driverId, _) = do
+  mbReferralCodeWithDriver <- B.runInReplica $ QRD.findById driverId
+
+  case mbReferralCodeWithDriver of
+    Just driverReferral -> pure $ GenerateReferralCodeRes driverReferral.referralCode.getId
+    Nothing -> do
+      Redis.whenWithLockRedis makeLastRefferalCodeKey 60 $ do
+        refferalCodeNumber <- CQD.getLastRefferalCode
+        let referralCode' = T.pack $ formatReferralCode (show (refferalCodeNumber + 1))
+        driverRefferalRecord <- mkDriverRefferalType referralCode'
+        void (QRD.create driverRefferalRecord)
+
+      mbReferralRecord <- B.runInReplica $ QRD.findById driverId
```

Comment:
Why do this?

Reply:
added withLockRedisAndReturnValue function in shared kernel and used here